[CSS] Add @scope rule to RuleSet

mdubet opened the pull request:
#### 5d2ddd6e61126370c7da861fd5f2034818a7960c
<pre>
[CSS] Add @scope rule to RuleSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=264500">https://bugs.webkit.org/show_bug.cgi?id=264500</a>
<a href="https://rdar.apple.com/118181109">rdar://118181109</a>

Reviewed by Antti Koivisto.

This patch implements the mechanism to get the ancestor @scope rules (if any)
during element rule collection.

It also desactivates style sharing for scoped rules.

The actual matching is not implemented yet.

<a href="https://drafts.csswg.org/css-cascade-6/#scoped-styles">https://drafts.csswg.org/css-cascade-6/#scoped-styles</a>

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/scope-at-rule-expected.html: Added.
* LayoutTests/fast/css/scope-at-rule.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-name-defining-rules-expected.txt:
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::addRule):
(WebCore::Style::RuleSet::scopeRulesFor const):
* Source/WebCore/style/RuleSet.h:
(WebCore::Style::RuleSet::hasScopeRules const):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
(WebCore::Style::RuleSetBuilder::addStyleRuleWithSelectorList):
* Source/WebCore/style/RuleSetBuilder.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::hasScopeRules const):
* Source/WebCore/style/StyleScopeRuleSets.h:
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::resolve):

Canonical link: <a href="https://commits.webkit.org/271078@main">https://commits.webkit.org/271078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/464f95085acf3647d1830c660db83fa88286f758

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27251 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28499 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29474 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24933 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27719 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3289 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24766 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27513 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23386 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4091 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4200 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30113 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24881 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24798 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30378 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4236 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2383 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28315 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5702 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24118 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6571 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4695 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4616 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/RuleSet.h:
```diff
@@ -118,15 +120,20 @@ class RuleSet : public RefCounted<RuleSet> {
     bool hasContainerQueries() const { return !m_containerQueries.isEmpty(); }
     Vector<const CQ::ContainerQuery*> containerQueriesFor(const RuleData&) const;
 
+    bool hasScopeQueries() const { return !m_scope.isEmpty(); }
```
I don't think "scope query" (unlike "container query") is a spec term.  Maybe `hasScopes()` or `hasScopedRules()`?

mdubet replied:
What about `hasScopeRules` / `m_scopeRules` / ...? Just `scope` seems a bit too vague

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -578,6 +582,69 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
     return true;
 }
 
+bool ElementRuleCollector::scopeQueriesMatch(const RuleData& ruleData, const MatchRequest& matchRequest)
```
Let's leave the matching part (everything in ElementRuleCollector) out from the first patch and just focus on the RuletSet and RuleSetBuilder.

anttijk commented on Source/WebCore/style/RuleSet.cpp:
```diff
@@ -460,5 +458,25 @@ void RuleSet::shrinkToFit()
+        identifier = query.parent;
+    };
+
+    return queries;
```
It is very inefficient to make copies of CSSSelectorLists in the middle of style resolution. I think this could just return Vector<const StyleRuleScope*>.

mdubet replied:
`Vector<Ref<const StyleRuleScope>>` ?

anttijk replied:
probably better!

anttijk commented on Source/WebCore/style/RuleSet.h:
```diff
@@ -118,15 +120,20 @@ class RuleSet : public RefCounted<RuleSet> {
 
+    bool hasScopeQueries() const { return !m_scope.isEmpty(); }
+    using ScopeQuery = std::pair<CSSSelectorList, CSSSelectorList>;
+    Vector<ScopeQuery> scopeQueriesFor(const RuleData&) const;
```
Similarly for the names here.

anttijk commented on Source/WebCore/style/RuleSet.cpp:
```diff
@@ -113,28 +113,26 @@ static bool shouldHaveBucketForAttributeName(const CSSSelector& attributeSelecto
+    };
+    storeIdentifier(cascadeLayerIdentifier, m_cascadeLayerIdentifierForRulePosition);
+    storeIdentifier(containerQueryIdentifier, m_containerQueryIdentifierForRulePosition);
+    storeIdentifier(scopeQueryIdentifier, m_scopeIdentifierForRulePosition);
```
nice

anttijk commented on Source/WebCore/style/RuleSet.h:
```diff
@@ -202,6 +214,10 @@ class RuleSet : public RefCounted<RuleSet> {
     Vector<ContainerQueryIdentifier> m_containerQueryIdentifierForRulePosition;
 
+    // @scope
+    Vector<ScopeAndParent> m_scope;
```
m_scopes

mdubet committed:
[CSS] Add @scope rule to RuleSet
https://bugs.webkit.org/show_bug.cgi?id=264500
rdar://118181109

Reviewed by Antti Koivisto.

This patch implements the mechanism to get the ancestor @scope rules (if any)
during element rule collection.

It also desactivates style sharing for scoped rules.

The actual matching is not implemented yet.

https://drafts.csswg.org/css-cascade-6/#scoped-styles

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/scope-at-rule-expected.html: Added.
* LayoutTests/fast/css/scope-at-rule.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-name-defining-rules-expected.txt:
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::addRule):
(WebCore::Style::RuleSet::scopeRulesFor const):
* Source/WebCore/style/RuleSet.h:
(WebCore::Style::RuleSet::hasScopeRules const):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
(WebCore::Style::RuleSetBuilder::addStyleRuleWithSelectorList):
* Source/WebCore/style/RuleSetBuilder.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::hasScopeRules const):
* Source/WebCore/style/StyleScopeRuleSets.h:
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::resolve):

Canonical link: https://commits.webkit.org/271078@main
